File: Sources/UIPalette.swift
import Foundation

#if canImport(AppKit)
import AppKit
#endif

#if canImport(SwiftUI)
import SwiftUI
#endif

enum UIPaletteID: String, CaseIterable, Identifiable, Codable {
  case classic
  case nord
  case gruvbox
  case mono
  case megaDark
  case oledBlack
  case rainbow
  case custom

  var id: String { rawValue }

  var displayName: String {
    switch self {
    case .classic: return "Classic"
    case .nord: return "Nord"
    case .gruvbox: return "Gruvbox"
    case .mono: return "Mono"
    case .megaDark: return "Mega Dark"
    case .oledBlack: return "OLED Black"
    case .rainbow: return "Rainbow"
    case .custom: return "Custom"
    }
  }
}

struct RGBAColor: Codable, Equatable {
  var r: Double
  var g: Double
  var b: Double
  var a: Double

  init(r: Double, g: Double, b: Double, a: Double = 1.0) {
    func clamp(_ v: Double) -> Double { max(0.0, min(1.0, v)) }
    self.r = clamp(r)
    self.g = clamp(g)
    self.b = clamp(b)
    self.a = clamp(a)
  }

  static func rgb255(_ r: Int, _ g: Int, _ b: Int, _ a: Int = 255) -> RGBAColor {
    RGBAColor(
      r: Double(max(0, min(255, r))) / 255.0,
      g: Double(max(0, min(255, g))) / 255.0,
      b: Double(max(0, min(255, b))) / 255.0,
      a: Double(max(0, min(255, a))) / 255.0
    )
  }

#if canImport(SwiftUI)
  func toSwiftUIColor() -> Color {
    Color(.sRGB, red: r, green: g, blue: b, opacity: a)
  }

  static func fromSwiftUIColor(_ c: Color) -> RGBAColor? {
#if canImport(AppKit)
    let ns = NSColor(c)
    guard let srgb = ns.usingColorSpace(.sRGB) else { return nil }
    return RGBAColor(
      r: Double(srgb.redComponent),
      g: Double(srgb.greenComponent),
      b: Double(srgb.blueComponent),
      a: Double(srgb.alphaComponent)
    )
#else
    _ = c
    return nil
#endif
  }
#endif
}

struct UIPaletteData: Codable, Equatable {
  var windowGradientA: RGBAColor
  var windowGradientB: RGBAColor
  var windowGradientC: RGBAColor
  var windowRadialSpot: RGBAColor
  var railPanelAccent: RGBAColor
  var contextPanelAccent: RGBAColor

  var captureAccent: RGBAColor
  var uploadAccent: RGBAColor
  var workflowsAccent: RGBAColor
  var toolsAccent: RGBAColor
  var afterCaptureAccent: RGBAColor
  var afterUploadAccent: RGBAColor
  var destinationsAccent: RGBAColor
  var settingsAccent: RGBAColor
  var historyAccent: RGBAColor
}

struct UIPalette: Equatable {
  var id: UIPaletteID
  var displayName: String
  var data: UIPaletteData
}

enum UIPaletteCatalog {
  static let presets: [UIPalette] = [
    UIPalette(id: .classic, displayName: UIPaletteID.classic.displayName, data: classic),
    UIPalette(id: .nord, displayName: UIPaletteID.nord.displayName, data: nord),
    UIPalette(id: .gruvbox, displayName: UIPaletteID.gruvbox.displayName, data: gruvbox),
    UIPalette(id: .mono, displayName: UIPaletteID.mono.displayName, data: mono),
    UIPalette(id: .megaDark, displayName: UIPaletteID.megaDark.displayName, data: megaDark),
    UIPalette(id: .oledBlack, displayName: UIPaletteID.oledBlack.displayName, data: oledBlack),
    UIPalette(id: .rainbow, displayName: UIPaletteID.rainbow.displayName, data: rainbow),
  ]

  static func defaultCustomSeed() -> UIPaletteData {
    classic
  }

  static func effectivePalette(selected: UIPaletteID, custom: UIPaletteData) -> UIPalette {
    if selected == .custom {
      return UIPalette(id: .custom, displayName: UIPaletteID.custom.displayName, data: custom)
    }
    if let preset = presets.first(where: { $0.id == selected }) {
      return preset
    }
    return UIPalette(id: .classic, displayName: UIPaletteID.classic.displayName, data: classic)
  }

  private static let classic = UIPaletteData(
    windowGradientA: .rgb255(26, 115, 242),
    windowGradientB: .rgb255(64, 217, 179),
    windowGradientC: .rgb255(250, 199, 56),
    windowRadialSpot: .rgb255(242, 77, 89),
    railPanelAccent: .rgb255(26, 115, 242),
    contextPanelAccent: .rgb255(64, 217, 179),
    captureAccent: .rgb255(242, 140, 51),
    uploadAccent: .rgb255(38, 179, 217),
    workflowsAccent: .rgb255(64, 199, 115),
    toolsAccent: .rgb255(242, 77, 89),
    afterCaptureAccent: .rgb255(250, 199, 56),
    afterUploadAccent: .rgb255(89, 217, 179),
    destinationsAccent: .rgb255(64, 166, 140),
    settingsAccent: .rgb255(140, 153, 179),
    historyAccent: .rgb255(64, 115, 242)
  )

  private static let nord = UIPaletteData(
    windowGradientA: .rgb255(46, 52, 64),
    windowGradientB: .rgb255(94, 129, 172),
    windowGradientC: .rgb255(163, 190, 140),
    windowRadialSpot: .rgb255(191, 97, 106),
    railPanelAccent: .rgb255(94, 129, 172),
    contextPanelAccent: .rgb255(136, 192, 208),
    captureAccent: .rgb255(208, 135, 112),
    uploadAccent: .rgb255(136, 192, 208),
    workflowsAccent: .rgb255(163, 190, 140),
    toolsAccent: .rgb255(191, 97, 106),
    afterCaptureAccent: .rgb255(235, 203, 139),
    afterUploadAccent: .rgb255(143, 188, 187),
    destinationsAccent: .rgb255(129, 161, 193),
    settingsAccent: .rgb255(229, 233, 240),
    historyAccent: .rgb255(94, 129, 172)
  )

  private static let gruvbox = UIPaletteData(
    windowGradientA: .rgb255(251, 73, 52),
    windowGradientB: .rgb255(184, 187, 38),
    windowGradientC: .rgb255(250, 189, 47),
    windowRadialSpot: .rgb255(211, 134, 155),
    railPanelAccent: .rgb255(250, 189, 47),
    contextPanelAccent: .rgb255(131, 165, 152),
    captureAccent: .rgb255(251, 73, 52),
    uploadAccent: .rgb255(131, 165, 152),
    workflowsAccent: .rgb255(184, 187, 38),
    toolsAccent: .rgb255(211, 134, 155),
    afterCaptureAccent: .rgb255(250, 189, 47),
    afterUploadAccent: .rgb255(142, 192, 124),
    destinationsAccent: .rgb255(254, 128, 25),
    settingsAccent: .rgb255(213, 196, 161),
    historyAccent: .rgb255(131, 165, 152)
  )

  private static let mono = UIPaletteData(
    windowGradientA: .rgb255(90, 90, 92),
    windowGradientB: .rgb255(140, 140, 145),
    windowGradientC: .rgb255(200, 200, 205),
    windowRadialSpot: .rgb255(120, 120, 125),
    railPanelAccent: .rgb255(110, 110, 115),
    contextPanelAccent: .rgb255(150, 150, 155),
    captureAccent: .rgb255(210, 210, 214),
    uploadAccent: .rgb255(190, 190, 194),
    workflowsAccent: .rgb255(170, 170, 175),
    toolsAccent: .rgb255(155, 155, 160),
    afterCaptureAccent: .rgb255(200, 200, 205),
    afterUploadAccent: .rgb255(180, 180, 185),
    destinationsAccent: .rgb255(160, 160, 165),
    settingsAccent: .rgb255(140, 140, 145),
    historyAccent: .rgb255(120, 120, 125)
  )

  private static let megaDark = UIPaletteData(
    windowGradientA: .rgb255(2, 3, 6),
    windowGradientB: .rgb255(6, 9, 15),
    windowGradientC: .rgb255(10, 14, 22),
    windowRadialSpot: .rgb255(24, 18, 42),
    railPanelAccent: .rgb255(36, 40, 58),
    contextPanelAccent: .rgb255(42, 52, 74),
    captureAccent: .rgb255(255, 106, 61),
    uploadAccent: .rgb255(0, 230, 255),
    workflowsAccent: .rgb255(97, 255, 170),
    toolsAccent: .rgb255(255, 64, 184),
    afterCaptureAccent: .rgb255(255, 214, 64),
    afterUploadAccent: .rgb255(118, 114, 255),
    destinationsAccent: .rgb255(189, 111, 255),
    settingsAccent: .rgb255(150, 161, 186),
    historyAccent: .rgb255(76, 180, 255)
  )

  private static let oledBlack = UIPaletteData(
    windowGradientA: .rgb255(0, 0, 0),
    windowGradientB: .rgb255(0, 0, 0),
    windowGradientC: .rgb255(0, 0, 0),
    windowRadialSpot: .rgb255(0, 0, 0),
    railPanelAccent: .rgb255(20, 24, 28),
    contextPanelAccent: .rgb255(18, 22, 26),
    captureAccent: .rgb255(255, 126, 65),
    uploadAccent: .rgb255(45, 212, 255),
    workflowsAccent: .rgb255(70, 255, 174),
    toolsAccent: .rgb255(255, 78, 174),
    afterCaptureAccent: .rgb255(255, 219, 82),
    afterUploadAccent: .rgb255(134, 125, 255),
    destinationsAccent: .rgb255(198, 118, 255),
    settingsAccent: .rgb255(185, 195, 214),
    historyAccent: .rgb255(92, 188, 255)
  )

  private static let rainbow = UIPaletteData(
    windowGradientA: .rgb255(255, 64, 64),
    windowGradientB: .rgb255(255, 184, 48),
    windowGradientC: .rgb255(76, 220, 120),
    windowRadialSpot: .rgb255(122, 92, 255),
    railPanelAccent: .rgb255(255, 64, 64),
    contextPanelAccent: .rgb255(76, 220, 120),
    captureAccent: .rgb255(255, 106, 61),
    uploadAccent: .rgb255(63, 197, 255),
    workflowsAccent: .rgb255(92, 235, 156),
    toolsAccent: .rgb255(255, 76, 187),
    afterCaptureAccent: .rgb255(255, 214, 64),
    afterUploadAccent: .rgb255(97, 113, 255),
    destinationsAccent: .rgb255(177, 92, 255),
    settingsAccent: .rgb255(147, 160, 255),
    historyAccent: .rgb255(0, 215, 255)
  )
}

File: Windows/CraftyCannon.Core/UiPalette.cs
using System.Globalization;
using System.Text.Json.Serialization;

namespace CraftyCannon.Core;

public sealed record RgbaColor(double Red, double Green, double Blue, double Alpha = 1.0)
{
    public static RgbaColor Rgb255(int red, int green, int blue, double alpha = 1.0) =>
        new(red / 255.0, green / 255.0, blue / 255.0, alpha);

    [JsonIgnore]
    public RgbaColor Normalized => new(
        ClampColor(Red),
        ClampColor(Green),
        ClampColor(Blue),
        ClampColor(double.IsNaN(Alpha) ? 1.0 : Alpha));

    public string ToHexRgba()
    {
        var color = Normalized;
        return string.Create(CultureInfo.InvariantCulture, $"#{ToByte(color.Red):X2}{ToByte(color.Green):X2}{ToByte(color.Blue):X2}{ToByte(color.Alpha):X2}");
    }

    public static bool TryParseHex(string? value, out RgbaColor color)
    {
        color = Rgb255(0, 0, 0);
        if (string.IsNullOrWhiteSpace(value))
        {
            return false;
        }

        var text = value.Trim();
        if (text.StartsWith('#'))
        {
            text = text[1..];
        }

        if (text.Length is not (6 or 8) || !int.TryParse(text, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var packed))
        {
            return false;
        }

        var red = text.Length == 6 ? (packed >> 16) & 0xFF : (packed >> 24) & 0xFF;
        var green = text.Length == 6 ? (packed >> 8) & 0xFF : (packed >> 16) & 0xFF;
        var blue = text.Length == 6 ? packed & 0xFF : (packed >> 8) & 0xFF;
        var alpha = text.Length == 6 ? 0xFF : packed & 0xFF;
        color = Rgb255(red, green, blue, alpha / 255.0).Normalized;
        return true;
    }

    private static double ClampColor(double value) => Math.Clamp(double.IsNaN(value) ? 0 : value, 0, 1);

    private static int ToByte(double value) => (int)Math.Round(Math.Clamp(value, 0, 1) * 255, MidpointRounding.AwayFromZero);
}

public sealed record UiPaletteData(
    RgbaColor WindowGradientA,
    RgbaColor WindowGradientB,
    RgbaColor WindowGradientC,
    RgbaColor WindowRadialSpot,
    RgbaColor RailPanelAccent,
    RgbaColor ContextPanelAccent,
    RgbaColor CaptureAccent,
    RgbaColor UploadAccent,
    RgbaColor WorkflowsAccent,
    RgbaColor ToolsAccent,
    RgbaColor AfterCaptureAccent,
    RgbaColor AfterUploadAccent,
    RgbaColor DestinationsAccent,
    RgbaColor SettingsAccent,
    RgbaColor HistoryAccent)
{
    public UiPaletteData Normalized(UiPaletteData? fallback = null)
    {
        var seed = fallback ?? UiPaletteCatalog.DefaultCustomSeed();
        return new UiPaletteData(
            NormalizeColor(WindowGradientA, seed.WindowGradientA),
            NormalizeColor(WindowGradientB, seed.WindowGradientB),
            NormalizeColor(WindowGradientC, seed.WindowGradientC),
            NormalizeColor(WindowRadialSpot, seed.WindowRadialSpot),
            NormalizeColor(RailPanelAccent, seed.RailPanelAccent),
            NormalizeColor(ContextPanelAccent, seed.ContextPanelAccent),
            NormalizeColor(CaptureAccent, seed.CaptureAccent),
            NormalizeColor(UploadAccent, seed.UploadAccent),
            NormalizeColor(WorkflowsAccent, seed.WorkflowsAccent),
            NormalizeColor(ToolsAccent, seed.ToolsAccent),
            NormalizeColor(AfterCaptureAccent, seed.AfterCaptureAccent),
            NormalizeColor(AfterUploadAccent, seed.AfterUploadAccent),
            NormalizeColor(DestinationsAccent, seed.DestinationsAccent),
            NormalizeColor(SettingsAccent, seed.SettingsAccent),
            NormalizeColor(HistoryAccent, seed.HistoryAccent));
    }

    private static RgbaColor NormalizeColor(RgbaColor? value, RgbaColor fallback) => (value ?? fallback).Normalized;
}

public sealed record UiPalette(string Id, string DisplayName, UiPaletteData Data);

public static class UiPaletteCatalog
{
    public const string ClassicId = "classic";
    public const string CustomId = "custom";

    public static IReadOnlyList<UiPalette> Presets =>
    [
        new(ClassicId, "Classic", Classic),
        new("nord", "Nord", Nord),
        new("gruvbox", "Gruvbox", Gruvbox),
        new("mono", "Mono", Mono),
        new("megaDark", "Mega Dark", MegaDark),
        new("oledBlack", "OLED Black", OledBlack),
        new("rainbow", "Rainbow", Rainbow)
    ];

    public static UiPaletteData DefaultCustomSeed() => Classic;

    public static string NormalizeId(string? selected)
    {
        if (string.Equals(selected, CustomId, StringComparison.OrdinalIgnoreCase))
        {
            return CustomId;
        }

        return Presets.Any(preset => string.Equals(preset.Id, selected, StringComparison.OrdinalIgnoreCase))
            ? Presets.First(preset => string.Equals(preset.Id, selected, StringComparison.OrdinalIgnoreCase)).Id
            : ClassicId;
    }

    public static UiPalette Resolve(string? selected, UiPaletteData? custom)
    {
        var id = NormalizeId(selected);
        if (id == CustomId)
        {
            return new UiPalette(CustomId, "Custom", (custom ?? DefaultCustomSeed()).Normalized(DefaultCustomSeed()));
        }

        return Presets.FirstOrDefault(preset => string.Equals(preset.Id, id, StringComparison.OrdinalIgnoreCase)) ?? Presets[0];
    }

    private static readonly UiPaletteData Classic = new(
        RgbaColor.Rgb255(26, 115, 242),
        RgbaColor.Rgb255(64, 217, 179),
        RgbaColor.Rgb255(250, 199, 56),
        RgbaColor.Rgb255(242, 77, 89),
        RgbaColor.Rgb255(26, 115, 242),
        RgbaColor.Rgb255(64, 217, 179),
        RgbaColor.Rgb255(242, 140, 51),
        RgbaColor.Rgb255(38, 179, 217),
        RgbaColor.Rgb255(64, 199, 115),
        RgbaColor.Rgb255(242, 77, 89),
        RgbaColor.Rgb255(250, 199, 56),
        RgbaColor.Rgb255(89, 217, 179),
        RgbaColor.Rgb255(64, 166, 140),
        RgbaColor.Rgb255(140, 153, 179),
        RgbaColor.Rgb255(64, 115, 242));

    private static readonly UiPaletteData Nord = new(
        RgbaColor.Rgb255(46, 52, 64), RgbaColor.Rgb255(94, 129, 172), RgbaColor.Rgb255(163, 190, 140), RgbaColor.Rgb255(191, 97, 106), RgbaColor.Rgb255(94, 129, 172), RgbaColor.Rgb255(136, 192, 208), RgbaColor.Rgb255(208, 135, 112), RgbaColor.Rgb255(136, 192, 208), RgbaColor.Rgb255(163, 190, 140), RgbaColor.Rgb255(191, 97, 106), RgbaColor.Rgb255(235, 203, 139), RgbaColor.Rgb255(143, 188, 187), RgbaColor.Rgb255(129, 161, 193), RgbaColor.Rgb255(229, 233, 240), RgbaColor.Rgb255(94, 129, 172));

    private static readonly UiPaletteData Gruvbox = new(
        RgbaColor.Rgb255(251, 73, 52), RgbaColor.Rgb255(184, 187, 38), RgbaColor.Rgb255(250, 189, 47), RgbaColor.Rgb255(211, 134, 155), RgbaColor.Rgb255(250, 189, 47), RgbaColor.Rgb255(131, 165, 152), RgbaColor.Rgb255(251, 73, 52), RgbaColor.Rgb255(131, 165, 152), RgbaColor.Rgb255(184, 187, 38), RgbaColor.Rgb255(211, 134, 155), RgbaColor.Rgb255(250, 189, 47), RgbaColor.Rgb255(142, 192, 124), RgbaColor.Rgb255(254, 128, 25), RgbaColor.Rgb255(213, 196, 161), RgbaColor.Rgb255(131, 165, 152));

    private static readonly UiPaletteData Mono = new(
        RgbaColor.Rgb255(90, 90, 92), RgbaColor.Rgb255(140, 140, 145), RgbaColor.Rgb255(200, 200, 205), RgbaColor.Rgb255(120, 120, 125), RgbaColor.Rgb255(110, 110, 115), RgbaColor.Rgb255(150, 150, 155), RgbaColor.Rgb255(210, 210, 214), RgbaColor.Rgb255(190, 190, 194), RgbaColor.Rgb255(170, 170, 175), RgbaColor.Rgb255(155, 155, 160), RgbaColor.Rgb255(200, 200, 205), RgbaColor.Rgb255(180, 180, 185), RgbaColor.Rgb255(160, 160, 165), RgbaColor.Rgb255(140, 140, 145), RgbaColor.Rgb255(120, 120, 125));

    private static readonly UiPaletteData MegaDark = new(
        RgbaColor.Rgb255(2, 3, 6), RgbaColor.Rgb255(6, 9, 15), RgbaColor.Rgb255(10, 14, 22), RgbaColor.Rgb255(24, 18, 42), RgbaColor.Rgb255(36, 40, 58), RgbaColor.Rgb255(42, 52, 74), RgbaColor.Rgb255(255, 106, 61), RgbaColor.Rgb255(0, 230, 255), RgbaColor.Rgb255(97, 255, 170), RgbaColor.Rgb255(255, 64, 184), RgbaColor.Rgb255(255, 214, 64), RgbaColor.Rgb255(118, 114, 255), RgbaColor.Rgb255(189, 111, 255), RgbaColor.Rgb255(150, 161, 186), RgbaColor.Rgb255(76, 180, 255));

    private static readonly UiPaletteData OledBlack = new(
        RgbaColor.Rgb255(0, 0, 0), RgbaColor.Rgb255(0, 0, 0), RgbaColor.Rgb255(0, 0, 0), RgbaColor.Rgb255(0, 0, 0), RgbaColor.Rgb255(20, 24, 28), RgbaColor.Rgb255(18, 22, 26), RgbaColor.Rgb255(255, 126, 65), RgbaColor.Rgb255(45, 212, 255), RgbaColor.Rgb255(70, 255, 174), RgbaColor.Rgb255(255, 78, 174), RgbaColor.Rgb255(255, 219, 82), RgbaColor.Rgb255(134, 125, 255), RgbaColor.Rgb255(198, 118, 255), RgbaColor.Rgb255(185, 195, 214), RgbaColor.Rgb255(92, 188, 255));

    private static readonly UiPaletteData Rainbow = new(
        RgbaColor.Rgb255(255, 64, 64), RgbaColor.Rgb255(255, 184, 48), RgbaColor.Rgb255(76, 220, 120), RgbaColor.Rgb255(122, 92, 255), RgbaColor.Rgb255(255, 64, 64), RgbaColor.Rgb255(76, 220, 120), RgbaColor.Rgb255(255, 106, 61), RgbaColor.Rgb255(63, 197, 255), RgbaColor.Rgb255(92, 235, 156), RgbaColor.Rgb255(255, 76, 187), RgbaColor.Rgb255(255, 214, 64), RgbaColor.Rgb255(97, 113, 255), RgbaColor.Rgb255(177, 92, 255), RgbaColor.Rgb255(147, 160, 255), RgbaColor.Rgb255(0, 215, 255));
}
